### PR TITLE
Cleanup ubuntu/16

### PR DESCRIPTION
**Pull request by sdmaclea**

### CODE REVIEW

File: src/ubuntu/16.04/Dockerfile [hunk 1]
```diff
@@ -11,11 +11,9 @@ RUN apt-get update && \
     apt-get update && \
     apt-get install -y \
             cmake \
-            clang-3.5 \
             clang-3.9 \
-            liblldb-3.9-dev && \
+            liblldb-3.9-dev \
```
Comment (sdmaclea) [inline src/ubuntu/16.04/Dockerfile]: fix typo

File: src/ubuntu/16.04/Dockerfile [hunk 2]
```diff
@@ -11,11 +11,9 @@ RUN apt-get update && \
     apt-get update && \
     apt-get install -y \
             cmake \
-            clang-3.5 \
```
Comment (sdmaclea) [inline src/ubuntu/16.04/Dockerfile]: obsolete

File: src/ubuntu/16.04/Dockerfile [hunk 3]
```diff
@@ -11,11 +11,9 @@ RUN apt-get update && \
...
             clang-3.9 \
-            liblldb-3.9-dev && \
+            liblldb-3.9-dev \
             lldb-3.9 \
-            llvm-3.5 \
```
Comment (sdmaclea) [inline src/ubuntu/16.04/Dockerfile]: obsolete

File: src/ubuntu/16.04/crossdeps/Dockerfile [hunk 4]
```diff
@@ -2,20 +2,14 @@ FROM microsoft/dotnet-buildtools-prereqs:ubuntu-16.04-coredeps
 
 # Install the base toolchain we need to build anything (clang, cmake, make and the like).
 RUN apt-get update \
-    && apt-get install -y wget \
-    && echo "deb http://llvm.org/apt/trusty/ llvm-toolchain-trusty-3.6 main" | tee /etc/apt/sources.list.d/llvm.list \
-    && wget -O - http://llvm.org/apt/llvm-snapshot.gpg.key | apt-key add - \
-    && apt-get update \
```
Comment (sdmaclea) [inline src/ubuntu/16.04/crossdeps/Dockerfile]: trusty packages shouldn't be installed in xenial

File: src/ubuntu/16.04/crossdeps/Dockerfile [hunk 5]
```diff
@@ -2,20 +2,14 @@ FROM microsoft/dotnet-buildtools-prereqs:ubuntu-16.04-coredeps
...
     && apt-get install -y \
-        binfmt-support \
         binutils-arm-linux-gnueabihf \
-        clang-3.6 \
```
Comment (sdmaclea) [inline src/ubuntu/16.04/crossdeps/Dockerfile]: obsolete

File: src/ubuntu/16.04/crossdeps/Dockerfile [hunk 6]
```diff
@@ -2,20 +2,14 @@ FROM microsoft/dotnet-buildtools-prereqs:ubuntu-16.04-coredeps
...
         cmake \
         debootstrap \
-        llvm-3.6 \
```
Comment (sdmaclea) [inline src/ubuntu/16.04/crossdeps/Dockerfile]: obsolete

File: src/ubuntu/16.04/crossdeps/Dockerfile [hunk 7]
```diff
@@ -2,20 +2,14 @@ FROM microsoft/dotnet-buildtools-prereqs:ubuntu-16.04-coredeps
...
         make \
         qemu \
         qemu-user-static \
+        wget \
```
Comment (sdmaclea) [inline src/ubuntu/16.04/crossdeps/Dockerfile]: Moved from above

File: src/ubuntu/16.04/crossdeps/Dockerfile [hunk 8]
```diff
@@ -2,20 +2,14 @@ FROM microsoft/dotnet-buildtools-prereqs:ubuntu-16.04-coredeps
 
 # Install the base toolchain we need to build anything (clang, cmake, make and the like).
 RUN apt-get update \
-    && apt-get install -y wget \
-    && echo "deb http://llvm.org/apt/trusty/ llvm-toolchain-trusty-3.6 main" | tee /etc/apt/sources.list.d/llvm.list \
-    && wget -O - http://llvm.org/apt/llvm-snapshot.gpg.key | apt-key add - \
-    && apt-get update \
     && apt-get install -y \
-        binfmt-support \
```
Comment (sdmaclea) [inline src/ubuntu/16.04/crossdeps/Dockerfile]: deferred to post-build.  This need to be run privileged.

File: src/ubuntu/build-scripts/build-rootfs.sh [hunk 9]
```diff
@@ -27,5 +27,6 @@ do
     buildRootFSContainer="rootfs-$arch-$crossToolset"
     docker run --privileged --rm --name $buildRootFSContainer -e ROOTFS_DIR=/rootfs/$arch \
         -v $PWD/rootfs:/rootfs -v $scriptsVolume:/scripts \
-        $dockerCrossDepsTag /scripts/cross/build-rootfs.sh $arch $crossToolset $lldb --skipunmount
+        $dockerCrossDepsTag /scripts/cross/build-rootfs.sh $arch $crossToolset $lldb --skipunmount || \
+        exit 1
```
Comment (sdmaclea) [inline src/ubuntu/build-scripts/build-rootfs.sh]: exit on error

**Comment by sdmaclea** [comment]
@jashook @janvorli @MichaelSimons This is an attempt ...  I'll look carefully at the build logs to see if this resolves the issues.

File: src/ubuntu/build-scripts/build-install-binfmt-support.sh [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+
+# Stop script on NZEC
+set -e
+# Stop script if unbound variable found (use ${var:-} if intentional)
+set -u
+
+os=$1
+dockerCrossDepsTag="microsoft/dotnet-buildtools-prereqs:${os}-crossdeps"
+
+docker run --privileged --read-only=false --rm $dockerCrossDepsTag apt-get install -y binfmt-support
```
Comment (MichaelSimons) [inline src/ubuntu/build-scripts/build-install-binfmt-support.sh]: Can you explain this?  I don't see how the resulting modification (e.g. installing binfmt-support) is being captured.

File: src/ubuntu/14.04/crossdeps/hooks/post-build [hunk 11]
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env bash
+
+$(dirname ${BASH_SOURCE[0]})/../../../build-scripts/build-install-binfmt-support.sh ubuntu-16.04
```
Comment (MichaelSimons) [inline src/ubuntu/14.04/crossdeps/hooks/post-build]: Last arg looks incorrect - shouldn't it be 14.04

Comment (sdmaclea) [inline src/ubuntu/14.04/crossdeps/hooks/post-build]: Oops.

File: src/ubuntu/build-scripts/build-install-binfmt-support.sh (hunk 10 above)
Comment (sdmaclea) [reply src/ubuntu/build-scripts/build-install-binfmt-support.sh]: By marking the image as `--read-only=false` I am assuming the modifications to the image are written back to the original image.

Comment (sdmaclea) [reply src/ubuntu/build-scripts/build-install-binfmt-support.sh]: Based on the logs, looks like this didn't fix anything.

File: src/ubuntu/14.04/crossdeps/hooks/post-build (hunk 11 above)
Comment (sdmaclea) [reply src/ubuntu/14.04/crossdeps/hooks/post-build]: Acknowledged

File: src/ubuntu/build-scripts/build-install-binfmt-support.sh (hunk 10 above)
Comment (MichaelSimons) [reply src/ubuntu/build-scripts/build-install-binfmt-support.sh]: No it is not.  A better approach would be to modify the build-rootfs script to start the container, `docker exec` to add `binfmt-support` and then `docker exec` to run `build-rootfs`

**Comment by sdmaclea** [comment]
I have trimmed this back to removing obsolete 16.04 packages and fixing the install script typo.

**Comment by MichaelSimons** [comment]
What about changes to ensure the build will fail when building the rootfs fails?  At a minimum this should get logged as an open issue.

**Comment by sdmaclea** [comment]
The addition of the `|| exit 1` also didn't work probably because the return code of the hook scripts are probably ignored by the builder.  I took it out because it didn't work.  #22 at least mentions the issue.

**Comment by MichaelSimons** [comment]
How certain are you that build-rootfs is returning non-zero in this case?  Non-zero return value from build hooks will fail the build.

**Comment by sdmaclea** [comment]
core-setup/cross/build-rootfs.sh is not exiting properly..  From the logs in #22 : 
`18:06:46 /scripts/cross/build-rootfs.sh: line 138: install_Failed: command not found`

Maybe because the function is defined after it is called.

**Comment by sdmaclea** [comment]
I'll play with fixing `core-setup/cross/build-rootfs.sh` tomorrow.

**Comment by sdmaclea** [comment]
Once that is fixed, I can make sure the hook script handles the error.

**Comment by sdmaclea** [comment]
test ubuntu_16

**Comment by sdmaclea** [comment]
@MichaelSimons The changes to core-setup now cause the rootfs build to fail.  See the ubuntu_16 logs.

**Comment by sdmaclea** [comment]
The `|| exit 1` might still be helpful for the multiple rootfs loop case.  I'll add it back.

**Comment by jashook** [comment]
test ubuntu_16

**Comment by sdmaclea** [comment]
test ubuntu_14

File: src/ubuntu/build-scripts/build-rootfs.sh (hunk 9 above)
Comment (MichaelSimons) [reply src/ubuntu/build-scripts/build-rootfs.sh]: @sdmaclea - would you mind adding a check to ensure $PWD/rootfs/$arch exists and is not empty after the docker run cmd?

Comment (jashook) [reply src/ubuntu/build-scripts/build-rootfs.sh]: opened #25 to address the issue

**Comment by MichaelSimons** [comment]
As I said in https://github.com/dotnet/dotnet-buildtools-prereqs-docker/issues/22#issuecomment-373229399, I still don't think things are working yet.  Notice this in the ubuntu_16 CI

> 19:32:22 -- EXECUTING: docker build -t microsoft/dotnet-buildtools-prereqs:ubuntu-16.04-cross-d809634-20180315012910 -f src/ubuntu/16.04/cross/Dockerfile src/ubuntu/16.04/cross
> 19:32:22 Sending build context to Docker daemon  8.192kB

I would expect the build context should be much larger if the rootfs folder is populated correctly.

**Comment by jashook** [comment]
https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/25 should fail if this is correct.

File: src/ubuntu/build-scripts/build-rootfs.sh (hunk 9 above)
Comment (MichaelSimons) [reply src/ubuntu/build-scripts/build-rootfs.sh]: The two PRs should get merged together so that we can run CI against the full set of changes we believe is necessary.

**Comment by sdmaclea** [comment]
The context sizes are all pretty small.  Not sure what that is, but it is not the size of the file system.

The arm64 context is smaller than the x86 + arm context

```
19:06:46 -- EXECUTING: docker build -t microsoft/dotnet-buildtools-prereqs:ubuntu-16.04-cross-arm64-d809634-20180315012910 -f src/ubuntu/16.04/cross/arm64/Dockerfile src/ubuntu/16.04/cross/arm64
19:06:46 Sending build context to Docker daemon  4.608kB
```

I tried to pull the docker tag, but I don't think it was actually published.

Also the `TESTING IMAGES` section is empty.  It possible test should really go there.

File: src/ubuntu/build-scripts/build-rootfs.sh (hunk 9 above)
Comment (sdmaclea) [reply src/ubuntu/build-scripts/build-rootfs.sh]: Done

**Comment by sdmaclea** [comment]
@MichaelSimons Looks like your assertion that things aren't quite working is correct.

**Comment by sdmaclea** [comment]
I just pushed an experimental patch to use `docker exec $buildRootFSContainer tar Ccf /rootfs - >> $PWD/rootfs.tar` to extract the rootfs image.

It won't work because `ADD` statements will need to be modified.  I'm also not sure the tar file paths will be correct for ADD.  So I just added a `tar -tf ` to list the tar ball contents

**Comment by MichaelSimons** [comment]
@sdmaclea - did you try `docker cp`?  I thought that would work but I ran into some permission issues.

```
    docker run --privileged --name $buildRootFSContainer -e ROOTFS_DIR=/rootfs/$arch \
        -v $scriptsVolume:/scripts \
        $dockerCrossDepsTag /scripts/cross/build-rootfs.sh $arch $crossToolset $lldb --skipunmount
    docker cp $buildRootFSContainer:/rootfs/. $PWD/rootfs/
```

> 21:20:12 -- EXECUTING: docker build -t microsoft/dotnet-buildtools-prereqs:ubuntu-16.04-cross-arm64-d809634-20180315034515 -f src/ubuntu/16.04/cross/arm64/Dockerfile src/ubuntu/16.04/cross/arm64
> 21:20:12 error checking context: 'no permission to read from '/repo/src/ubuntu/16.04/cross/arm64/rootfs/arm64/dev/agpgart''.
> 21:20:12 System.InvalidOperationException: Failed to execute docker build -t microsoft/dotnet-buildtools-prereqs:ubuntu-16.04-cross-arm64-d809634-20180315034515 -f src/ubuntu/16.04/cross/arm64/Dockerfile src/ubuntu/16.04/cross/arm64

**Comment by sdmaclea** [comment]
@MichaelSimons I did not try `docker cp` .  Based on reading the man page, I decided it would not work.  

In previous experience with these rootfs...  I had run tar as an ordinary user to move a rootfs to another machine, when I untarred it was unusable.  To get ti to work I had to tar it as the sudo user.  So I was thinking to get permissions and special files correct tar would be the right approach.

Why are we building it, then moving it to the destination image?  
Why don't we simply build it in the destination image?

**Comment by MichaelSimons** [comment]
IIRC in order to build the rootfs in a docker container, you need to be running in `–privileged` mode.  You cannot build with a Dockerfile in privileged mode.  Therefore a container was spun up to build the rootfs and then copied into the dest image.

**Comment by MichaelSimons** [comment]
@sdmaclea  - Given what you said, it be difficult to build the rootfs, push it to azure blob storage and then then simply download it in the Dockerfile?  Just trying to think of a quick way to unblock this.

**Comment by jashook** [comment]
I believe the approach of using tar then modifying the ADD statement should get us what we need.

**Comment by sdmaclea** [comment]
> Therefore a container was spun up to build the rootfs and then copied into the dest image.

I was wondering if it would be easier to do in a post-build hook in the destination image
+ Add a rootfs-install.sh script into destination image
+ Add a hook post-build script to run rootfs-install.sh in privileged mode
+ The rootfs-install.sh script would
  + git clone core-setup
  + run core-setup/cross/buildrootfs.sh with arguments
  + error check and cleanup... rm -rf core-setup

Seems simpler with fewer moving parts.

**Comment by sdmaclea** [comment]
The tar failed in the log because the container wasn't running.
I'll try to work through the steps locally then update the patch

These errors in the log concern me too.  Might need to be fixed in core-setup
```
03:40:53 dangling: /usr/lib/llvm-3.9/lib/liblldb-3.9.1.so -> liblldb-3.9.so
03:40:53 dangling: /usr/lib/llvm-3.9/lib/liblldb-3.9.so -> liblldb-3.9.so.1
03:40:53 dangling: /usr/lib/llvm-3.9/lib/liblldb.so -> liblldb-3.9.so
```
Opened https://github.com/dotnet/core-setup/issues/3839

**Comment by sdmaclea** [comment]
Pushed a revised patch which I am hopeful will do the right thing.

**Comment by sdmaclea** [comment]
Ubuntu_16 build failed, debugged, revised patch...  

>  it be difficult to build the rootfs, push it to azure blob storage and then then simply download it in the Dockerfile?

I have never used azure blob storage, but I think this should work.

I think the patch is close, but if you want to abandon this that would be a viable alternative.  It still might not unblock things quicker because someone would need to create the tar balls for azure.

**Comment by sdmaclea** [comment]
test debian

Failed to fetch debian stretch packages

**Comment by jashook** [comment]
Thanks for the work Steve, fingers crossed on the rootfs build. I believe the changes look good. If not they seem very close.

**Comment by sdmaclea** [comment]
Ubuntu_16 cross-arm64 completed
`13:44:49 Sending build context to Docker daemon  1.078GB`

**Comment by sdmaclea** [comment]
debian build failed with same error as earlier.
```
14:20:48 Get:12 https://packages.microsoft.com/repos/microsoft-debian-stretch-prod stretch/main amd64 Packages [9656 B]
14:20:48 Err:12 https://packages.microsoft.com/repos/microsoft-debian-stretch-prod stretch/main amd64 Packages
14:20:48   Writing more data than expected (9832 > 9656)
14:20:48   Hashes of expected file:
14:20:48    - Filesize:9656 [weak]
14:20:48    - SHA512:2ee32046003228b4809f6b4efe58399d3d6d3bfc36519e57f7b2577cb600ab840c41613d04b3159099018e279647f83ea0d51f098f2997832936e48744219cf3
14:20:48    - SHA256:48fc4d95ebe26eccffd5ebb29137bfcc0c1d7e2a3f16dd205d3d94694d10d535
14:20:48    - SHA1:dc785f1756cd4905d5e783dcfdb70e29f41a2c1b [weak]
14:20:48    - MD5Sum:57a36478eed4d5611380ed6a03415e50 [weak]
14:20:48   Release file created at: Tue, 13 Mar 2018 19:49:07 +0000
14:20:49 Fetched 10.3 MB in 1s (7825 kB/s)
14:20:49 Reading package lists...
14:20:49 [91mE: Failed to fetch https://packages.microsoft.com/repos/microsoft-debian-stretch-prod/dists/stretch/main/binary-amd64/Packages.gz  Writing more data than 
```

**Comment by sdmaclea** [comment]
Everything else has passed.  I think this can be merged?

**Comment by sdmaclea** [comment]
Looks like rootfs images could be shrunk substantially by adding

`rm -rf /rootfs/$arch/var/cache/apt/archives/*` 
and/or
`rm -rf /rootfs/$arch/var/lib/apt/lists/*`

prior to tarring

This could be added into #23, but since #23 is blocking other work perhaps this should be deferred and done with #16

**Comment by MichaelSimons** [comment]
The following ubuntu/16.04 images were built as a result of this PR:

microsoft/dotnet-buildtools-prereqs:ubuntu-14.04-a3ae44b-20180316024003
microsoft/dotnet-buildtools-prereqs:ubuntu-14.04-debpkg-d770b8b-20180316024003
microsoft/dotnet-buildtools-prereqs:ubuntu-14.04-coredeps-d770b8b-20180316024003
microsoft/dotnet-buildtools-prereqs:ubuntu-14.04-crossdeps-d770b8b-20180316024003
microsoft/dotnet-buildtools-prereqs:ubuntu-14.04-cross-a3ae44b-20180316024003
microsoft/dotnet-buildtools-prereqs:ubuntu-16.04-a3ae44b-20180316023254
microsoft/dotnet-buildtools-prereqs:ubuntu-16.04-debpkg-d770b8b-20180316023254
microsoft/dotnet-buildtools-prereqs:ubuntu-16.04-coredeps-d770b8b-20180316023254
microsoft/dotnet-buildtools-prereqs:ubuntu-16.04-crossdeps-620ba25-20180316023254
microsoft/dotnet-buildtools-prereqs:ubuntu-16.04-cross-arm64-a3ae44b-20180316023254
microsoft/dotnet-buildtools-prereqs:ubuntu-16.04-cross-a3ae44b-20180316023254
